File: crypto_agent/tools/analyze.py
from .registry import register_tool
from ..indicators import compute_sma, compute_rsi, compute_bollinger, generate_signals


@register_tool(
    name="analyze",
    description="Technical analysis for a symbol: SMA, RSI, Bollinger Bands, and buy/sell signals.",
    schema={
        "type": "object",
        "properties": {
            "symbol": {"type": "string", "description": "Trading pair, e.g. BTC/USDT", "default": "BTC/USDT"},
            "timeframe": {"type": "string", "description": "Candle interval", "default": "1h"},
        },
        "required": [],
    },
)
async def handle_analyze(exchange, symbol: str = "BTC/USDT", timeframe: str = "1h", **_) -> str:
    try:
        klines = await exchange.fetch_ohlcv(symbol, timeframe, 100)
        closes = [k["close"] if isinstance(k, dict) else k[4] for k in klines]

        if len(closes) < 2:
            return f"Insufficient data for {symbol} ({len(closes)} candles)."

        last_price = closes[-1]
        sma20 = compute_sma(closes, 20)
        sma50 = compute_sma(closes, 50)
        rsi = compute_rsi(closes)
        bb_lower, bb_mid, bb_upper = compute_bollinger(closes)

        lines = [
            f"Technical Analysis: {symbol} ({timeframe})",
            f"{'=' * 45}",
            f"Last Price: {last_price:.2f}",
            "",
            "Moving Averages:",
            f"  SMA(20):  {sma20:.2f}" + (" ⚠️ insufficient data" if sma20 == 0 else ""),
            f"  SMA(50):  {sma50:.2f}" + (" ⚠️ insufficient data" if sma50 == 0 else ""),
            "",
            f"RSI(14):    {rsi:.1f}" + (" [Oversold]" if rsi < 30 else " [Overbought]" if rsi > 70 else " [Neutral]"),
            "",
            "Bollinger Bands(20, 2):",
            f"  Upper:  {bb_upper:.2f}",
            f"  Middle: {bb_mid:.2f}",
            f"  Lower:  {bb_lower:.2f}",
        ]

        if sma20 and sma50:
            trend = "BULLISH" if sma20 > sma50 else "BEARISH"
            lines.extend(["", f"Trend: {trend} (SMA20 {'>' if sma20 > sma50 else '<'} SMA50)"])

        signals = generate_signals(closes, last_price)
        lines.extend(["", f"Trading Signals:", f"{'=' * 45}"])
        for sig in signals:
            lines.append(f"  • {sig}")

        buy_count = sum(1 for s in signals if "BUY" in s or "BULLISH" in s)
        sell_count = sum(1 for s in signals if "SELL" in s or "BEARISH" in s)
        if buy_count > sell_count:
            lines.append(f"\nOverall: BULLISH ({buy_count} buy vs {sell_count} sell)")
        elif sell_count > buy_count:
            lines.append(f"\nOverall: BEARISH ({sell_count} sell vs {buy_count} buy)")
        else:
            lines.append("\nOverall: NEUTRAL")

        return "\n".join(lines)
    except Exception as e:
        return f"Error in analyze: {e}"

File: src/tools/analyze.ts
import { registerTool } from "./registry.js";
import { computeSma, computeRsi, computeBollinger, generateSignals } from "../indicators.js";

registerTool(
  "analyze",
  "Technical analysis for a symbol: SMA, RSI, Bollinger Bands, and buy/sell signals.",
  {
    type: "object",
    properties: {
      symbol: { type: "string", description: "Trading pair, e.g. BTC/USDT", default: "BTC/USDT" },
      timeframe: { type: "string", description: "Candle interval", default: "1h" },
    },
    required: [],
  },
  async ({ exchange, symbol = "BTC/USDT", timeframe = "1h" }) => {
    try {
      const klines = await exchange.fetchOhlcv(symbol, timeframe, 100);
      const closes: number[] = klines.map((k: any) => (typeof k === "object" ? k.close : k[4]));

      if (closes.length < 2) return `Insufficient data for ${symbol} (${closes.length} candles).`;

      const lastPrice = closes[closes.length - 1];
      const sma20 = computeSma(closes, 20);
      const sma50 = computeSma(closes, 50);
      const rsi = computeRsi(closes);
      const [bbLower, bbMid, bbUpper] = computeBollinger(closes);

      const lines = [
        `Technical Analysis: ${symbol} (${timeframe})`,
        "=".repeat(45),
        `Last Price: ${lastPrice.toFixed(2)}`,
        "",
        "Moving Averages:",
        `  SMA(20):  ${sma20.toFixed(2)}${sma20 === 0 ? " ⚠️ insufficient data" : ""}`,
        `  SMA(50):  ${sma50.toFixed(2)}${sma50 === 0 ? " ⚠️ insufficient data" : ""}`,
        "",
        `RSI(14):    ${rsi.toFixed(1)}${rsi < 30 ? " [Oversold]" : rsi > 70 ? " [Overbought]" : " [Neutral]"}`,
        "",
        "Bollinger Bands(20, 2):",
        `  Upper:  ${bbUpper.toFixed(2)}`,
        `  Middle: ${bbMid.toFixed(2)}`,
        `  Lower:  ${bbLower.toFixed(2)}`,
      ];

      if (sma20 && sma50) {
        const trend = sma20 > sma50 ? "BULLISH" : "BEARISH";
        lines.push("", `Trend: ${trend} (SMA20 ${sma20 > sma50 ? ">" : "<"} SMA50)`);
      }

      const signals = generateSignals(closes, lastPrice);
      lines.push("", "Trading Signals:", "=".repeat(45));
      for (const sig of signals) lines.push(`  \u2022 ${sig}`);

      const buyCount = signals.filter((s) => s.includes("BUY") || s.includes("BULLISH")).length;
      const sellCount = signals.filter((s) => s.includes("SELL") || s.includes("BEARISH")).length;
      if (buyCount > sellCount) lines.push(`\nOverall: BULLISH (${buyCount} buy vs ${sellCount} sell)`);
      else if (sellCount > buyCount) lines.push(`\nOverall: BEARISH (${sellCount} sell vs ${buyCount} buy)`);
      else lines.push("\nOverall: NEUTRAL");

      return lines.join("\n");
    } catch (e: any) {
      return `Error in analyze: ${e.message ?? e}`;
    }
  },
);
